Add tag classes

eslickj opened the pull request:
## Summary/Motivation:

This adds a couple classes that let you tag quantities in a model for reporting output and setting input.  There are two purposes 1) easily identify important quantities in the complex model hierarchy and 2) allow a user to specify how the quantities should be displayed when writing to an SVG or table, or other report.

The is motivated by two things, I write similar features as a one-off in every flowsheet I put together for the PFD output, and I'm spending a lot of time searching all the time for the numbers I need.  Then when I hand off the model, it's hard to describe to someone what the key variables are.   I hope that this will simplify things. 


### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

andrewlee94 reviewed (changes requested):
As much as I hate to add more work (and be predictable), we will need documentation for this. AS you have commented you spend a lot of time doing this, so it is certain other people are doing the same thus it will be important to let them know that tools exist to do this already (and explain how to use it).

eslickj commented:
@andrewlee94, thanks I'll add some docs.

dallan-keylogic commented on idaes/core/util/tags.py:
```diff
@@ -0,0 +1,432 @@
+        return self.expression
+
+
+class ModelTagGroup(dict):
```
Several methods in this class need docstrings.

eslickj commented on idaes/core/util/tags.py:
```diff
@@ -0,0 +1,432 @@
+                else:
+                    raise KeyError(f"{index} is not a valid index for tag {self._name}")
+        else:
+            e = self.expression
```
That's right.  There a default index that gets passed in, whether it is indexed or not from some places.  It's a little weird, but it simplifies dealing with time.  I want to ignore that if it's not indexed.  This will let me group time indexed and not time indexed quantities together.

eslickj commented:
@andrewlee94 and @dallan-keylogic, I fixed some problems and added docs.  I also removed the idea of a default index for display.  It's was weird and confusing.  I some cases It'd be nice to have flat scalar tags, for like display on a PFD.  I think the best approach would be to have sort of flatting method that returns a new flat group.  Like say I have a dynamic model, and I want a group of tags with just the start time or end time.  I'll work that out later.  There are probably a couple small PRs to come to make this go more seamlessly with the SVG output.

dallan-keylogic commented on docs/technical_specs/core/util/tags.rst:
```diff
@@ -0,0 +1,109 @@
+      m.f = pyo.Expression(expr=m.x[1] / m.y)
+
+      @m.Expression([1, 2, 3], ["a", "b"])
+      def g(b, i, j):
```
Suggestion: change `g` to `h` here or `g` to `group` below, to avoid having both `g` and `m.g`.

andrewlee94 commented on docs/technical_specs/core/util/tags.rst:
```diff
@@ -0,0 +1,109 @@
+  m = model()
+  tag = ModelTag(expr=m.w, format_string="{:.3f}", display_units=pyo.units.g)
+
+Now we an use the tag to set model input and display model output.
```
Typo: can

andrewlee94 commented on docs/technical_specs/core/util/tags.rst:
```diff
@@ -0,0 +1,109 @@
+===============
+
+IDAES contains classes for tagging model quantities and grouping them.  The tags
+allow a convenient short cut to important model inputs and outputs.  There are
```
Minor comment, but I would combine the last two statements into a full sentence for better flow.

andrewlee94 commented on docs/technical_specs/core/util/tags.rst:
```diff
@@ -0,0 +1,109 @@
+
+The next code snippet creates a single tag object for the model variable ``w``.
+While the structure of the example model is simple, IDAES models often have a
+complex structure. Tags generally provide a much shorter means to reference a
```
Another minor comment: I would replace the final "." with ", and " to combine the final statement in to the previous sentence for better flow.

andrewlee94 reviewed (changes requested):
One typo and a couple of minor suggestions to improve the flow of the docs.
